Firmware/MotorControl/commands.c

```diff
@@ -335,21 +344,34 @@ void cmd_parse_thread(void const * argument) {
                     }
                 }
             }
-            // When we reach here, we are out of immediate characters to fetch out of UART buffer
-            // Now we check if there is any USB processing to do: we wait for up to 1 ms,
-            // before going back to checking UART again.
-            int USB_check_timeout = 1;
-            // Wait for signalling from USB interrupt (OTG_FS_IRQHandler)
-            osStatus semaphore_status = osSemaphoreWait(sem_usb_irq, USB_check_timeout);
-            if (semaphore_status == osOK) {
-                // We have a new incoming USB transmission: handle it
-                HAL_PCD_IRQHandler(&hpcd_USB_OTG_FS);
-                // Let the irq (OTG_FS_IRQHandler) fire again.
-                HAL_NVIC_EnableIRQ(OTG_FS_IRQn);
+            osStatus sem_stat = osSemaphoreWait(sem_usb_rx, 1);
+            if(sem_stat == osOK){
+                motor_parse_cmd(usb_buf, usb_len, SERIAL_PRINTF_IS_USB);
+                USBD_CDC_ReceivePacket(&hUsbDeviceFS); // Allow next packet
             }
         } while (!reset_read_state);
     }
-
     // If we get here, then this task is done
     vTaskDelete(osThreadGetId());
+}
+
+// Called from CDC_Receive_FS callback function, this allows motor_parse_cmd to access the
+// incoming USB data
+void set_cmd_buffer(const uint8_t *buf, uint32_t len) {
+    memcpy(usb_buf, buf, len);
```

Comment:
We don't really need to do this copy, as we won't be recieving any new data to overwrite what's there until we have processed it. We can instead have the set_cmd_buffer just be setting a pointer to `UserRxBufferFS` instead of taking a copy.